envoy/ssl/handshaker.h

```diff
@@ -156,5 +175,90 @@ class HandshakerFactory : public Config::TypedFactory {
   virtual SslCtxCb sslctxCb(HandshakerFactoryContext& handshaker_factory_context) const PURE;
 };
 
+struct SelectionResult {
+  enum class SelectionStatus {
+    // Continue the TLS handshake.
+    Continue,
+    // Block the TLS handshake.
+    Stop,
+    // Abort the TLS handshake.
+    AbortHandshake,
+  };
+  // If the value is Pending, the selection is asynchronous.
+  SelectionStatus status;
+  // It will be nullptr when status is Stop or AbortHandshake.
+  const Ssl::TlsContext* selected_ctx;
+  // Enable OCSP stapling response when it is true.
+  bool staple;
+};
+
+/**
+ * Used to return the result from an synchronous/asynchronous cert selection.
+ */
+class CertSelectionCallback {
+public:
+  virtual ~CertSelectionCallback() = default;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
```

Comment:
Hm. I such a thread needs a dispatcher, can we pass it in explicitly? It's really a odd fit for this interface. Where is this method called in this PR?

Reply:
yep, I agree, it is a odd fit.

It's used in the test case, to mock selecting cert in async mode.
https://github.com/envoyproxy/envoy/blob/1a85885ebe40d17409e9b48a03160bbb43e1461c/test/common/tls/tls_certificate_selector_test.cc#L100

It's is not required, when the extension is pure C++ and without new thread, but it could be useful when we implement an Golang extension, since Go code might run in another Go thread, not the Envoy work thread.

> can we pass it in explicitly?

Do you mean pass it as an argument, right?
`selectTlsContext(const SSL_CLIENT_HELLO& ssl_client_hello, CertificateSelectionCallbackPtr cb, Event::Dispatcher& dispatcher)`

but the `dispatcher` won't be used in most cases, so I prefer to hide it into the `CertificateSelectionCallback`.
otherwise, people may be confused about the `dispatcher`.

I'm glad to update it, if you think pass it as an argument could be better.

Reply:
If that's the only place we use the dispatcher, could we modify TestTlsCertificateSelectorFactory so that it takes a dispatcher in the constructor and just pass it in in the test?

Reply:
Okay, removed, we can figure out a way, when we need it in the feature.